python/pyarrow/array.pxi

```diff
@@ -803,45 +806,60 @@ cdef class Date64Value(ArrayValue):
             ap.Value(self.index) / 1000).date()
 
 
+cdef dict DATETIME_CONVERSION_FUNCTIONS
+
+try:
+    import pandas as pd
+except ImportError:
+    DATETIME_CONVERSION_FUNCTIONS = {
+        TimeUnit_SECOND: lambda x, tzinfo: (
+            datetime.datetime.utcfromtimestamp(x).replace(tzinfo=tzinfo)
+        ),
+        TimeUnit_MILLI: lambda x, tzinfo: (
+            datetime.datetime.utcfromtimestamp(x / 1e3).replace(tzinfo=tzinfo)
+        ),
+        TimeUnit_MICRO: lambda x, tzinfo: (
+            datetime.datetime.utcfromtimestamp(x / 1e6).replace(tzinfo=tzinfo)
+        ),
+    }
+else:
+    DATETIME_CONVERSION_FUNCTIONS = {
+        TimeUnit_SECOND: lambda x, tzinfo: pd.Timestamp(
+            x * 1000000000, tz=tzinfo, unit='ns',
+        ),
+        TimeUnit_MILLI: lambda x, tzinfo: pd.Timestamp(
+            x * 1000000, tz=tzinfo, unit='ns',
+        ),
+        TimeUnit_MICRO: lambda x, tzinfo: pd.Timestamp(
+            x * 1000, tz=tzinfo, unit='ns',
+        ),
+        TimeUnit_NANO: lambda x, tzinfo: pd.Timestamp(
+            x, tz=tzinfo, unit='ns',
+        )
+    }
```

Comment:
This is pretty awkward to factor out because of the `TimeUnit_*` enum values. We'd have to make `pandas_compat.pxi` if we wanted to keep those available to Cython but not Python (which would seem to defeat part of the purpose of factoring out) or expose the enum values to Python. This doesn't seem worth it for something that will never be seen by a user. Still, if you feel strongly about it I can spend some more time on it.